### ISSUE TITLE
revalidatePath Not Working

### ISSUE THREAD

**Issue by gchallen** [caching-tools/next-shared-cache#275]
**Brief Description of the Bug**

Calling `revalidatePath` does not remove cached data from the Redis cache and force a rebuild of an ISR page.

**Severity**

This seems like it should work.

**Frequency of Occurrence**

Always.

**Steps to Reproduce**
Provide detailed steps to reproduce the behavior, including any specific conditions or configurations where the bug occurs:

1. Set up a Redis-only cache
2. Load a page generated using ISR
3. Verify that the page is cached by Redis
4. Call `revalidatePath` in a route handler, for example in a `POST` method intended for on-demand revalidation
5. Verify that the Redis cache still contains the key that should be removed
6. Reload the page and view the stale data

**Expected vs. Actual Behavior**

`revalidatePath` should remove the cached data and force the page to be rebuilt.

**Environment:**

-   OS: Kubernetes running Linux pods in production
-   18.19.0
-   0.6.6
-   14.1.0

**Attempted Solutions or Workarounds**

Creating a Redis client and manually deleting the offending keys in the revalidation `POST` method seems to work.

**Impact of the Bug**

Again, this seems like something that should work automatically. It's possible that I've misconfigured something, but my cache configuration is fairly simple.

**Comment by better-salmon** [comment]
Hello @gchallen! Thank you for opening this issue. I am convinced you configured everything well. I'll try to explain.

When you call the `revalidatePath` or `revalidateTag` method from the `next/cache` module, it does not remove any values from the cache store. It doesn't happen either in the original cache or my cache. These functions add tags and prefixed paths to a `revalidatedTags` list, which also lives in the cache store. When Next.js retrieves the data, it also retrieves `revalidatedTags`. Then, my cache does some logic with the tags baked in the cache value and tags from the `revalidatedTags` list and decides if the page or data is staled.

If your page is updated after calling the `revalidatePath` or `revalidateTag`, it works as intended. If this is the case, let's move this issue to discussions. I will check if I can implement the logic you are expecting. But I need more time.

**Comment by gchallen** [comment]
Thanks for the explanation. I did poke around the code a bit and noticed this logic maintaining the `revalidatedTags` list, but didn't quite understand how it worked.

My use case is more or less what's described [here](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#using-on-demand-revalidation): performing programmatic on-demand revalidation. Initially I was using the `res.revalidate` method as shown, but I found that that was fairly slow, and seemed to be actually waiting for the page to rebuild. (I could probably just skip the `await` I guess.) Given that my endpoint is sometimes revalidating a bunch of pages, I moved to the `revalidatePath` approach which was much faster—but it's more clear now why...

I don't fully understand the NextJS caching logic, but it does seem like there should be some way to force a page rebuild on the _next_ load, without immediately triggering a page rebuild, which can cause an unnecessary load spike when revalidating many pages, particularly if some of them are infrequently accessed. Manually deleting the page from the Redis cache works for my use case, but only because I've disabled the in-memory caches entirely.
